lib/dockerregistry/checks.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package inventory
+
+import (
+	"encoding/hex"
+	"fmt"
+	"os"
+	"strings"
+
+	gogit "gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
+
+// ImageRemovalCheck implements the PreCheck interface and checks against
+// pull requests that attempt to remove any images from the promoter manifests.
+type ImageRemovalCheck struct {
+	GitRepoPath    string
+	MasterSHA      plumbing.Hash
+	PullRequestSHA plumbing.Hash
+	PromotionEdges map[PromotionEdge]interface{}
+}
+
+func getGitShaFromEnv(envVar string) (plumbing.Hash, error) {
+	potenitalSHA := os.Getenv(envVar)
+	const gitShaLength = 40
+	if len(potenitalSHA) != gitShaLength {
+		return plumbing.Hash{},
+			fmt.Errorf("Length of SHA is %d characters, should be %d",
+				len(potenitalSHA), gitShaLength)
+	}
+	_, err := hex.DecodeString(potenitalSHA)
+	if err != nil {
+		return plumbing.Hash{}, fmt.Errorf("Not a valid SHA: %v", err)
+	}
+	return plumbing.NewHash(potenitalSHA), nil
+}
+
+// MKRealImageRemovalCheck returns an instance of ImageRemovalCheck.
+func MKRealImageRemovalCheck(
+	gitRepoPath string,
+	edges map[PromotionEdge]interface{},
+) (*ImageRemovalCheck, error) {
+	masterSHA, err := getGitShaFromEnv("PULL_BASE_SHA")
+	if err != nil {
+		return nil, fmt.Errorf("The PULL_BASE_SHA environment variable "+
+			"is invalid: %v", err)
+	}
+	pullRequestSHA, err := getGitShaFromEnv("PULL_PULL_SHA")
+	if err != nil {
+		return nil, fmt.Errorf("The PULL_PULL_SHA environment variable "+
+			"is invalid: %v", err)
+	}
+	return &ImageRemovalCheck{
+		gitRepoPath,
+		masterSHA,
+		pullRequestSHA,
+		edges,
+	}, nil
+}
+
+// Run executes ImageRemovalCheck on a set of promotion edges.
+// Returns an error if the pull request removes images from the
+// promoter manifests.
+func (check *ImageRemovalCheck) Run() error {
+	var err error
+	// To compare the master and pull request promoter manifests, check
```

Comment:
Ping? I commented here but it looks like my old comment is gone. Basically, state *why* in comments, not so much the *how*. The *what*-type of comments here is the least useful.

Reply:
done